File: Data Structures & Algorithms/partition-equal-subset-sum/submission-4.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int x : nums)
            sum+=x;
        if(sum%2!=0) return false;
        sum /= 2;
        boolean[][] dp = new boolean[nums.length][sum];
        return helper(nums, 0, 0, sum, dp);
    }
    public boolean helper(int[] nums, int i, int curr, int sum, boolean[][] dp) {
        if(curr==sum) {
            return true;
        }
        if(curr>sum || i==nums.length)
            return false;
        if(dp[i][curr])
            return dp[i][curr];
        dp[i][curr] = helper(nums, i+1, curr+nums[i], sum, dp) || helper(nums, i+1, curr, sum, dp);
        return dp[i][curr];
    }
}

File: Data Structures & Algorithms/target-sum/submission-4.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        S = sum(nums)
        dp = [[-1]*(2*S +1) for _ in range(n)]

        def helper(i,curr):
            if i==n:
                return 1 if curr==target else 0
            if dp[i][curr+S]!=-1:
                return dp[i][curr+S]
            ret = helper(i+1,curr+nums[i]) + helper(i+1,curr-nums[i])
            dp[i][curr+S] = ret
            return ret

        return helper(0,0)